In [1]:
import pandas as pd 
df = pd.read_csv("lululemon_feature_matrix.csv")
df.head()

keyhole  10"  wool-knit  cpc  side-snap  rest  hiking  unisex  v  \
0        0    0          0    0          0     0       0       0  0   
1        0    0          0    0          0     0       0       0  0   
2        0    0          0    0          0     0       0       0  0   
3        0    0          0    0          0     0       0       0  0   
4        0    0          0    0          0     0       0       0  0   

   button-up  ...  this  cord  train  tapered  textured  a-d  super-high-rise  \
0          0  ...     0     0      0        0         0    0                0   
1          0  ...     0     0      0        0         0    0                0   
2          0  ...     0     0      0        0         0    0                0   
3          0  ...     0     0      0        0         0    0                0   
4          0  ...     0     0      0        0         0    0                0   

   nulux  ruched  Price  
0      0       0  118.0  
1      0       0  128.0  
2      0       0   78.0  
3      0       0   78.0  
4      0       0  118.0  

[5 rows x 708 columns]

In [3]:
w_df_all_data = df[df['womens'] == 1]
m_df_all_data = df[df['womens'] == 0]

print(w_df_all_data.shape, m_df_all_data.shape)

(846, 708) (537, 708)


In [4]:
w_df_all_data.describe()

keyhole         10"  wool-knit         cpc   side-snap        rest  \
count  846.000000  846.000000      846.0  846.000000  846.000000  846.000000   
mean     0.002364    0.001182        0.0    0.005910    0.001182    0.001182   
std      0.048593    0.034381        0.0    0.076695    0.034381    0.034381   
min      0.000000    0.000000        0.0    0.000000    0.000000    0.000000   
25%      0.000000    0.000000        0.0    0.000000    0.000000    0.000000   
50%      0.000000    0.000000        0.0    0.000000    0.000000    0.000000   
75%      0.000000    0.000000        0.0    0.000000    0.000000    0.000000   
max      1.000000    1.000000        0.0    1.000000    1.000000    1.000000   

           hiking      unisex           v  button-up  ...        this  \
count  846.000000  846.000000  846.000000      846.0  ...  846.000000   
mean     0.009456    0.004728    0.002364        0.0  ...    0.009456   
std      0.096840    0.068639    0.048593        0.0  ...    0.096840   
min      0.000000    0.000000    0.000000        0.0  ...    0.000000   
25%      0.000000    0.000000    0.000000        0.0  ...    0.000000   
50%      0.000000    0.000000    0.000000        0.0  ...    0.000000   
75%      0.000000    0.000000    0.000000        0.0  ...    0.000000   
max      1.000000    1.000000    1.000000        0.0  ...    1.000000   

             cord       train  tapered  textured         a-d  super-high-rise  \
count  846.000000  846.000000    846.0     846.0  846.000000       846.000000   
mean     0.002364    0.073286      0.0       0.0    0.002364         0.013002   
std      0.048593    0.260760      0.0       0.0    0.048593         0.113351   
min      0.000000    0.000000      0.0       0.0    0.000000         0.000000   
25%      0.000000    0.000000      0.0       0.0    0.000000         0.000000   
50%      0.000000    0.000000      0.0       0.0    0.000000         0.000000   
75%      0.000000    0.000000      0.0       0.0    0.000000         0.000000   
max      1.000000    1.000000      0.0       0.0    1.000000         1.000000   

            nulux      ruched       Price  
count  846.000000  846.000000  846.000000  
mean     0.001182    0.002364   77.794917  
std      0.034381    0.048593   43.278870  
min      0.000000    0.000000    9.000000  
25%      0.000000    0.000000   48.000000  
50%      0.000000    0.000000   68.000000  
75%      0.000000    0.000000   99.000000  
max      1.000000    1.000000  398.000000  

[8 rows x 708 columns]

In [5]:
m_df_all_data.describe()

keyhole         10"   wool-knit         cpc  side-snap   rest  \
count    537.0  537.000000  537.000000  537.000000      537.0  537.0   
mean       0.0    0.011173    0.001862    0.009311        0.0    0.0   
std        0.0    0.105209    0.043153    0.096133        0.0    0.0   
min        0.0    0.000000    0.000000    0.000000        0.0    0.0   
25%        0.0    0.000000    0.000000    0.000000        0.0    0.0   
50%        0.0    0.000000    0.000000    0.000000        0.0    0.0   
75%        0.0    0.000000    0.000000    0.000000        0.0    0.0   
max        0.0    1.000000    1.000000    1.000000        0.0    0.0   

           hiking      unisex      v   button-up  ...   this   cord  \
count  537.000000  537.000000  537.0  537.000000  ...  537.0  537.0   
mean     0.007449    0.003724    0.0    0.003724  ...    0.0    0.0   
std      0.086065    0.060971    0.0    0.060971  ...    0.0    0.0   
min      0.000000    0.000000    0.0    0.000000  ...    0.0    0.0   
25%      0.000000    0.000000    0.0    0.000000  ...    0.0    0.0   
50%      0.000000    0.000000    0.0    0.000000  ...    0.0    0.0   
75%      0.000000    0.000000    0.0    0.000000  ...    0.0    0.0   
max      1.000000    1.000000    0.0    1.000000  ...    0.0    0.0   

            train     tapered   textured    a-d  super-high-rise       nulux  \
count  537.000000  537.000000  537.00000  537.0            537.0  537.000000   
mean     0.054004    0.005587    0.01676    0.0              0.0    0.001862   
std      0.226236    0.074604    0.12849    0.0              0.0    0.043153   
min      0.000000    0.000000    0.00000    0.0              0.0    0.000000   
25%      0.000000    0.000000    0.00000    0.0              0.0    0.000000   
50%      0.000000    0.000000    0.00000    0.0              0.0    0.000000   
75%      0.000000    0.000000    0.00000    0.0              0.0    0.000000   
max      1.000000    1.000000    1.00000    0.0              0.0    1.000000   

       ruched       Price  
count   537.0  537.000000  
mean      0.0   96.619181  
std       0.0   46.943974  
min       0.0   14.000000  
25%       0.0   59.000000  
50%       0.0   96.000000  
75%       0.0  128.000000  
max       0.0  369.000000  

[8 rows x 708 columns]

Once again, the median of the mens clothing is lower than the median of the womens clothing. This means that the lasso results will probably imply that "mens" actually increases prices whereas "womens" decreases it. But I wonder if random forest + SHAP values will have to say the same thing. 

In [6]:
print(df.isnull().sum().sum())
print(df.nunique())

0
keyhole             2
10"                 2
wool-knit           2
cpc                 2
side-snap           2
                   ..
a-d                 2
super-high-rise     2
nulux               2
ruched              2
Price              80
Length: 708, dtype: int64


In [7]:
correlations = df.corr()
print(correlations.loc['womens', 'Price'])

-0.201028350370316


A much more strongly negative correlation than Nike's, which was only -0.05, approximately. 

In [9]:
from sklearn.model_selection import train_test_split
x = df.iloc[:, :-1]
y = df.iloc[:, -1]
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state=0)
x_train.to_csv("x_train.csv", index=False)
x_test.to_csv("x_test.csv", index=False)
y_train.to_csv("y_train.csv", index=False)
y_test.to_csv("y_test.csv", index=False)

# save these files as csv so that we use the same data for all different types of models. 

X train shape:  (1037, 707) 
 Y train shape:  (1037, 1) 
 X test shape:  (346, 707) 
 Y test shape:  (346, 1)


In [3]:
x_train = pd.read_csv("x_train.csv")
x_test = pd.read_csv("x_test.csv")
y_train = pd.read_csv("y_train.csv")
y_test = pd.read_csv("y_test.csv")


print("X train shape: ", x_train.shape, 
      "\n Y train shape: ", y_train.shape,
       "\n X test shape: ", x_test.shape, 
         "\n Y test shape: ", y_test.shape)


X train shape:  (1037, 707) 
 Y train shape:  (1037, 1) 
 X test shape:  (346, 707) 
 Y test shape:  (346, 1)


In [4]:
#1. lasso (alpha = 1)
from sklearn.linear_model import Lasso
lasso = Lasso(alpha=1)
lasso.fit(x_train, y_train)
y_pred = lasso.predict(x_test)

In [5]:
# stats
from sklearn.metrics import mean_squared_error
from scipy.stats import pearsonr
print("MSE: ", mean_squared_error(y_test, y_pred))
y_test = y_test.values.flatten()
corr, _ = pearsonr(y_test, y_pred)
print("Correlation: ", corr)

MSE:  1479.819080610456
Correlation:  0.5185031966221738


In [6]:
# that's terrible, let's try a different alpha value. 

lasso = Lasso(alpha=0.01)
lasso.fit(x_train, y_train)
y_pred = lasso.predict(x_test)
print("MSE: ", mean_squared_error(y_test, y_pred))
#y_test = y_test.values.flatten()
corr, _ = pearsonr(y_test, y_pred)
print("Correlation: ", corr)

MSE:  675.5647720884425
Correlation:  0.8106082450183435


After playing around with different alpha values it appears that alpha=0.01 gives us the best results with the lowest MSE and highest correlation, going by alpha value to the 0.001. I'm sure there's a better way to find the ideal alpha value but I'll figure that out another day. 

Now let's see what these coefficients have to say. 

In [33]:
coefs = lasso.coef_
feature_names = x_train.columns
coefs_df = pd.DataFrame({'Feature': feature_names, 'Coefficient': coefs})
coefs_df = coefs_df[coefs_df['Coefficient'] != 0.0]
print(coefs_df)
#coefs_df.to_csv("coefs.csv", index=False)

             Feature  Coefficient
0            keyhole   -11.611843
1                10"   -14.097912
2          wool-knit    58.404503
3                cpc     7.954917
5               rest    10.976016
..               ...          ...
699             cord    -0.450938
700            train     0.475675
702         textured    38.852923
704  super-high-rise   -16.657445
706           ruched   -16.535639

[386 rows x 2 columns]


In [34]:
print(coefs_df.shape)

(386, 2)


So about half the words were found to be influential. that's not great. let's at least see what the "womens" coefficient was, along with what the top 10 most influential words were. 

In [35]:
coefs_df = coefs_df.sort_values(by='Coefficient', ascending=False)

In [41]:
print("These were the words in the product descriptions behind the biggest price increases, accroding to Lasso: \n")
print(coefs_df.iloc[0:10])

These were the words in the product descriptions behind the biggest price increases, accroding to Lasso: 

         Feature  Coefficient
297  down-filled   151.774012
523        parka   134.014472
678         puff   133.909356
400     overalls    93.478486
399   navigation    80.166956
565         coat    78.144616
281        rebel    77.652953
157          25l    77.411228
357       duffle    75.071296
567      another    74.331474


In [42]:
print("These were the words in the product descriptions behind the biggest price DECREASES, accroding to Lasso: \n")

print(coefs_df.iloc[-10:])

These were the words in the product descriptions behind the biggest price DECREASES, accroding to Lasso: 

             Feature  Coefficient
165              hat   -33.870869
299             clip   -36.278938
413         keychain   -37.413208
468         headband   -39.124689
355       cap-sleeve   -42.463554
506           velvet   -43.304241
352            socks   -49.368297
256  relaxed-tapered   -50.085643
599  classic-tapered   -75.750488
543       scrunchies   -76.097227


I think a general improvement for me would be to separate the adjectives from the nouns. Obviously socks are going to be cheaper than puffers. I think this would provide a better insight into the pricing models. That is for another week. Perhaps I should cut down on my number of datasets then and focus on just 3: Lululemon, Nike, and Gymshark. I think what I do with the data is more important than how many datasets I cover. I am doing a personal project, not a research paper. I should value depth over breadth. 

### to do: remove nouns and do an ADJECTIVE analysis!!! 

In [45]:
womens_coef = coefs_df[coefs_df['Feature'].str.contains('womens', case=False, na=False)]
print(womens_coef)

    Feature  Coefficient
631  womens     -1.13435


Yup, pretty low compared to the others (meaning it is not all that influential), and seems to be more favoring women than men. Now let's stop pretending that the pricing model is linear and let's apply some rfr! (Random forest regression)

In [7]:
from sklearn.ensemble import RandomForestRegressor
rfr = RandomForestRegressor(random_state=16)
rfr.fit(x_train, y_train)
y_pred = rfr.predict(x_test)

mse = mean_squared_error(y_test, y_pred)
corr, _ = pearsonr(y_test, y_pred)
print("Pearson correlation coefficient:", corr)
print("Mean squared error: ", mse)

/var/folders/f2/btr2krsx7hx6jcjm7m5dn87m0000gn/T/ipykernel_8042/2580542958.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rfr.fit(x_train, y_train)


Pearson correlation coefficient: 0.799649497675338
Mean squared error:  712.2667368030976


Wow it actually performed worse than lasso. Perhaps their pricing model is linear? Lol. I'm not sure it's worth it to sit down and look at the SHAP values right now. 

I think what I really need to do is just get rid of the nouns. They are messing everything up. We need adjectives only. I'm going to have to research how to do that. I think there are already pre-trained models that can easily filter out adjectives from nouns. Worst case, I can always just feed my dataset into chat gpt!!!! That'll filter out all the nouns for me. 

### Next steps: 
1. Remove nouns from Nike and Lululemon datasets and see how just plain adjectives influence price. If the adjectives improve the analyses, then yay! Don't bother with SHAP on lululemon for non-adjective dataset. Just rerun the whole thing on all the datasets but without adjectives (gymshark too) and finish up the project. 
2. If removal of adjectives doesn't improve anything, then run lasso and rfr on Gymshark. If the gymshark ML results show that womens "decreases" price, then you know something is wrong, and reach out to Abhi for help. 
3. However, if the gymshark ML results show that womens "increases" prices, then you know all your code is working. Just finish running SHAP values on lululemon price analysis (just to make everything even), and also run everything (lasso and rfr) on Gymshark data too.
4. Clean up code and publish. 

A suggestion I got was to try to reduce the number of features selected during each random forest step. Let's try it. 

In [23]:
rfr = RandomForestRegressor(max_features=30, random_state=16)
rfr.fit(x_train, y_train)
y_pred = rfr.predict(x_test)

mse = mean_squared_error(y_test, y_pred)
corr, _ = pearsonr(y_test, y_pred)
print("Pearson correlation coefficient:", corr)
print("Mean squared error: ", mse)

/var/folders/f2/btr2krsx7hx6jcjm7m5dn87m0000gn/T/ipykernel_8042/4110881440.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rfr.fit(x_train, y_train)


Pearson correlation coefficient: 0.8107210592487684
Mean squared error:  685.1503800740472


Okay yeah it didn't improve it all that much. The highest I could find was .81 correlation which is still not amazing compared to lasso. 

I think we should try removing the adjectives. 